src/uadk_prov_digest.c

```diff
@@ -158,19 +158,30 @@ static int uadk_digests_soft_md(struct digest_priv_ctx *priv)
 
 static int uadk_digest_soft_init(struct digest_priv_ctx *priv)
 {
-	return EVP_DigestInit(priv->soft_ctx, priv->soft_md);
+	if (priv->soft_md)
+		return EVP_DigestInit(priv->soft_ctx, priv->soft_md);
+
+	return 0;
 }
 
-static int uadk_digest_soft_update(EVP_MD_CTX *ctx, const void *data, size_t len)
+static int uadk_digest_soft_update(struct digest_priv_ctx *priv,
+				   const void *data, size_t len)
 {
-	return EVP_DigestUpdate(ctx, data, len);
+	if (priv->soft_md)
+		return EVP_DigestUpdate(priv->soft_ctx, data, len);
+
+	return 0;
 }
 
 static int uadk_digest_soft_final(struct digest_priv_ctx *priv, unsigned char *digest)
 {
-	unsigned int digest_length = EVP_MD_get_size(priv->soft_md);
+	if (priv->soft_md) {
+		unsigned int digest_length = EVP_MD_get_size(priv->soft_md);
+
+		return EVP_DigestFinal(priv->soft_ctx, digest, &digest_length);
```

Comment:
EVP_DigestFinal() is obsolete. Refer to OpenSSL document:
/* The functions EVP_DigestInit(), EVP_DigestFinal() and EVP_MD_CTX_copy() are
obsolete but are retained to maintain compatibility with existing code. New
applications should use EVP_DigestInit_ex(), EVP_DigestFinal_ex() and
EVP_MD_CTX_copy_ex() because they can efficiently reuse a digest context
instead of initializing and cleaning it up on each call and allow non default
implementations of digests to be specified. */
/*
=item EVP_DigestFinal()

Similar to EVP_DigestFinal_ex() except after computing the digest
the digest context I<ctx> is automatically cleaned up with EVP_MD_CTX_reset().
*/